### ISSUE TITLE
Ibus cycles through only the first two language engines.

### ISSUE THREAD

**Issue by mcognetta** [ibus/ibus#2303]
Please fill in the following items if you don't know the root cause.

Which distribution and version?: 
Ubuntu 20.04.2 LTS

Which desktop environment and version?: 
LXQt

Which session type?: 
X11

IBus version?: 
IBus 1.5.22

Issue description:
Ibus only switches between first two languages in engine list. For example, I have English, Esperanto, and Korean (ibus-hangul) in my ibus config. When switching using the next input method hot key, it will switch to the next language in the list, but when switching again, it goes back to the first one. If I manually select (through the task bar icon) my third language, it will then begin to cycle between the third language and the language that was most previously selected.

For example: on start up I have English open. I hit my toggle key, it switches to Esperanto. I hit toggle again it switches to English. I hit toggle again, it switches to Esperanto. I manually select Korean. I hit toggle, it switches to Esperanto. I hit toggle again, and it goes back to Korean.

Steps to reproduce:
1. Have 3+ languages installed.
2. Start iBus
3. Toggle to next input method 
4. Toggle again (should return to the original input method
5. Manually switch to a 3rd language.
6. Toggle, which should send you back to the first language
7. Toggle again, which should send you back to the _3rd_language.

Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
yes (actually this crashed my machine, yikes)

Do you see any errors when you run ibus-daemon with the verbose option?:
no

Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 
yes

----------------

I thought maybe that ibus was unable to detect the other languages, but I can observe them in my ibus config.

Here is my default config (I am posting this in case it provides some additional detail, I will post only the relevant parts after):

```
SCHEMA: org.freedesktop.ibus
SCHEMA: org.freedesktop.ibus.general
  enable-by-default: false
  xkb-latin-layouts: ['ara', 'bg', 'cz', 'dev', 'gr', 'gur', 'in', 'jp(kana)', 'mal', 'mkd', 'ru', 'ua']
  use-xmodmap: true
  preload-engines: ['xkb:us::eng', 'm17n:eo:x-sistemo', 'hangul']
  dconf-preserve-name-prefixes: ['/desktop/ibus/engine/pinyin', '/desktop/ibus/engine/bopomofo', '/desktop/ibus/engine/hangul']
  use-global-engine: true
  engines-order: ['xkb:us::eng', 'hangul', 'm17n:eo:x-sistemo']
  embed-preedit-text: true
  switcher-delay-time: 400
  use-system-keyboard-layout: false
  version: '1.5.22'
SCHEMA: org.freedesktop.ibus.general.hotkey
  previous-engine: @as []
  enable-unconditional: @as []
  disable-unconditional: @as []
  next-engine: ['Alt+Shift_L']
  trigger: ['Control+space', 'Zenkaku_Hankaku', 'Alt+Kanji', 'Alt+grave', 'Hangul', 'Alt+Release+Alt_R']
  next-engine-in-menu: ['Alt+Shift_L']
  prev-engine: @as []
  triggers: ['Super_L']
SCHEMA: org.freedesktop.ibus.panel
  show: 0
  show-im-name: false
  y: -1
  xkb-icon-rgba: '#415099'
  show-icon-on-systray: true
  auto-hide-timeout: 10000
  lookup-table-orientation: 1
  custom-font: 'Sans 10'
  use-custom-font: false
  property-icon-delay-time: 500
  follow-input-cursor-when-always-shown: false
  x: -1
SCHEMA: org.freedesktop.ibus.panel.emoji
  has-partial-match: false
  favorite-annotations: @as []
  load-unicode-at-startup: false
  partial-match-length: 3
  favorites: @as []
  hotkey: ['<Control><Shift>e']
  lang: 'en'
  font: 'Monospace 16'
  load-emoji-at-startup: true
  unicode-hotkey: ['<Control><Shift>u']
  partial-match-condition: 0
```

The relevant part is:

```
  engines-order: ['xkb:us::eng', 'hangul', 'm17n:eo:x-sistemo']
```

Upon toggling to another language, it changes to

```
  engines-order: ['hangul', 'xkb:us::eng', 'm17n:eo:x-sistemo']
```

And toggling again gives

```
  engines-order: ['xkb:us::eng', 'hangul', 'm17n:eo:x-sistemo']
```

Manually selecting Esperanto then gives

```
  engines-order: ['m17n:eo:x-sistemo', 'xkb:us::eng', 'hangul']
```

Toggling gives

```
  engines-order: ['xkb:us::eng', 'm17n:eo:x-sistemo', 'hangul']
```

and toggling again gives

```
  engines-order: ['m17n:eo:x-sistemo', 'xkb:us::eng', 'hangul']
```

As you can see, it is stuck in a loop between the first two elements of the engine list. I have tried this with more languages with the same effect.

**Comment by mcognetta** [comment]
I also get

```
** (ibus-engine-m17n:4648): CRITICAL **: 01:55:49.010: ibus_m17n_engine_callback: assertion 'm17n != NULL' failed
```

when running in verbose mode. It always happens when switching languages, but also it seems to happen at other times.

**Comment by mcognetta** [comment]
I did find a way around this, but it is still not ideal. My current hotkey is set to Super_L (windows key). I added another <Control><Space>. Now, if I hit control and tap space, a tab panel pops up on my screen allowing me to cycle through all the languages I have installed. 

It would still be nice to be able to cycle through them all just by hitting Super_L repeatedly though.

**Comment by jcguu95** [comment]
I have the switching key bound to `Shift+Space`. Pressing `Shift+Space` `Shift+Space` `Shift+Space` did follow what you described. However, `Shift+Space+Space+...` does allow it to go to deeper options.

**Comment by mcognetta** [comment]
Yes, this seems to be what I observed as well. I wonder if it is that single key-stroke toggles (like Super_L) do not allow it to go deeper while multikey ones do since you can hold the modifier keydown and hit the other repeatedly (in the case of Shift+Space like you mentioned).

**Comment by jcguu95** [comment]
Hm.. I think that key-stroke just replaces the current one with the selected one. So after the first attempt, `[_1_ 2 3 4..]` becomes `[_2_ 1 3 4]`. No wonder if you try that again, it gets you `1` instead of `3`. I'd say this behavior is fine.

**Comment by bytrangle** [comment]
I observed the same thing. `Ctrl + shift + space + space` does the trick for me: it will show a list of available inputs at the center of the screen, then I can press `space` to cycle through all the inputs.

Demo

https://github.com/user-attachments/assets/57c3796d-d3b3-4c97-8561-5d871e6e53ee

This behavior is fine, but I wish it was more obvious.

**Comment by fujiwarat** [comment]
If you have more than 2 ibus engines, IBus should show the popup.
If you wish to hide it, you could set `switcher-delay-time` property to `-1`

```
% gsettings get org.freedesktop.ibus.general switcher-delay-time
400
% gsettings set org.freedesktop.ibus.general switcher-delay-time -1
```
